### PR TITLE
Cleanup guestbook example resources

### LINKED ISSUE #16998: Clean up copies of individual resources in examples/guestbook
Follow-up #14256. See https://github.com/kubernetes/kubernetes/pull/14256#issuecomment-154583601

cc @linzichang @bgrant0607

### PR DESCRIPTION
**Pull request by paralin**
This patch separates the guestbook example files from the test resources used in various tests through the codebase. Fixes #16998 

cc @janetkuo @bgrant0607

### CODE REVIEW

**Comment by paralin** [comment]
"No space left on device" build failure oh boy

**Comment by janetkuo** [comment]
Was trying to assign this to myself on my phone...

**Comment by bgrant0607** [comment]
@paralin Thanks. The main point of the tests is to ensure that the example itself is tested. Ideally the tests would exercise the exact configurations used by the example.

@jeffmendoza @amygdala

**Comment by paralin** [comment]
@bgrant0607 This is probably a good middle ground. I think adding all-in-one would be good for the tests.

**Comment by janetkuo** [comment]
@k8s-bot ok to test

**Comment by janetkuo** [comment]
cc @ixdy: @k8s-bot isn't responding

**Comment by alex-mohr** [comment]
@k8s-bot test this please

**Comment by alex-mohr** [comment]
@janetkuo unsure if it was the string or you're not in the bot's list or it was a one-off glitch, but testing now underway

**Comment by paralin** [comment]
@janetkuo @alex-mohr Is there any way for me to see what failed in the Jenkins unit/integration tests?

**Comment by ixdy** [comment]
@paralin if you take the test artifacts link for the e2e job and remove everything after the git commit sha, you can navigate to the test results for the unit/integration test job - i.e. go to https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/add46641e1b85a23bc7ef34d433e8cda7b875b20/ and navigate from there.

**Comment by paralin** [comment]
```
	examples_test.go:403: ../examples/guestbook/frontend.yaml: frontend does not have a test case defined
	examples_test.go:403: ../examples/guestbook/redis-slave.yaml: redis-slave does not have a test case defined
	examples_test.go:435: Directory ../examples/guestbook: Expected 6 examples, Got 2
```

Looking into it...

**Comment by paralin** [comment]
@janetkuo @bgrant0607 Issue is that the tests expect 1 resource = 1 file. In `example_test.go`. As frontend.yaml contains two, it's annoying to set this up. 

There are two options to fix this:

 - Make a subdir named test-data and separate frontend.yaml into two files, duplicate all of the master-controller.yaml for redis. This has the downside that you're not actually testing the same file users would use.
 - Separate frontend.yaml into two files and update the docs to look like this again. Downside: undo some of what the recent docs changes have done.

I'm going to try to separate frontend into two files, put it into test-data, and see if putting a relative path into the filename in the test set will work properly. It should. This is kind of a middle ground between these two solutions.

**Comment by paralin** [comment]
The jenkins tests wouldn't like having a yaml file that isn't tested as all the yaml files are walked over in the code. Deferring to option #1 for now to fix the tests in hopes some of the code in examples_test could be refactored later to allow multiple resources in a file.

**Comment by bgrant0607** [comment]
The examples test needs to be fixed to handle multiple resources per file.

**Comment by janetkuo** [comment]
As mentioned in https://github.com/kubernetes/kubernetes/issues/16998#issuecomment-156255896:
> - The following single resource files should be removed from `examples/guestbook`:
  - `frontend-controller.yaml`
  - `frontend-service.yaml`
  - `redis-slave-controller.yaml`
  - `redis-slave-service.yaml`
- Some tests (*excluding* example tests), such as `test-cmd.sh` or `apply_test.go`, use the single resource example files that need to be removed from `examples/guestbook`. So we won't delete those single resource files but move them to `hack/testdata/guestbook` instead. 
- Example tests (`examples/examples_test.go`) should test *all* files under `examples/`. 
- Example tests need to be fixed to be able to test example files with multiple resources. 

You don't need examples/guestbook/test-data/, and examples_test.go should test all files under examples/. It needs to first be fixed to test yaml files containing multiple resources.

**Comment by paralin** [comment]
@janetkuo I'm aware, I was trying to make a workaround to keep the same examples_test.go and still test the definitions. I'll look into adjusting the go test file

**Comment by bgrant0607** [comment]
cc @jeffmendoza

**Comment by bgrant0607** [comment]
ok to test

File: examples/examples_test.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func TestExampleObjectSchemas(t *testing.T) {
 		"../cmd/integration": {
 			"v1-controller": &api.ReplicationController{},
 		},
-		"../examples/guestbook": {
+		"../examples/guestbook/test-data": {
```
Comment (bgrant0607) [inline examples/examples_test.go]: We want the examples test to test the configs actually used by the example. That's why the test exists. 

We need to extend it to handle multiple resources in one file.

**Comment by ixdy** [comment]
I'm considering this PR stale and closing it. Please reopen if you are still working on it.

**Pull request closed without merging**